File: fui_system/src/platform/qt/qt_wrapper/cpp/qbytearray.cpp
#include <QByteArray>
#include "qbytearray.h"

void *QByteArray_new()
{
    return static_cast<void *>(new (std::nothrow) QByteArray());
}

void QByteArray_delete(void *self)
{
    delete static_cast<QByteArray *>(self);
}

const char *QByteArray_constData(void *self)
{
    QByteArray *qByteArray = static_cast<QByteArray *>(self);
    return qByteArray->constData();
}

int QByteArray_size(void *self)
{
    QByteArray *qByteArray = static_cast<QByteArray *>(self);
    return qByteArray->size();
}

File: fui_system/src/platform/qt/qt_wrapper/qbytearray.rs
use crate::FUISystemError;

pub struct QByteArray {
    pub this: *mut ::std::os::raw::c_void,
}

impl QByteArray {
    #[allow(dead_code)]
    pub fn new() -> Result<Self, FUISystemError> {
        unsafe {
            let this = crate::platform::qt::qt_wrapper::QByteArray_new();
            if this.is_null() {
                return Err(FUISystemError::OutOfMemory);
            }

            Ok(Self { this })
        }
    }

    pub fn as_bytes(&self) -> &[u8] {
        unsafe {
            let ptr = crate::platform::qt::qt_wrapper::QByteArray_constData(self.this);
            let length = crate::platform::qt::qt_wrapper::QByteArray_size(self.this);
            std::slice::from_raw_parts(ptr as *const u8, length as usize)
        }
    }
}

impl Drop for QByteArray {
    fn drop(&mut self) {
        unsafe {
            crate::platform::qt::qt_wrapper::QByteArray_delete(self.this);
        }
    }
}
